File: BubbleSort/Swift/Sources/BubbleSort/BubbleSort.swift
//
//  BubbleSort.swift
//  Created by Ardalan Samimi on 2018-05-29.
//
//	Bubble sort is a comparision-based sorting algorithm, where each
//	element is compared to an adjacent element and swapped if they're
//	out of order.
//
//  Worst-case complexity: O(n^2), where n is the number of elements
//

func bubbleSort<T: Comparable>(_ array: [T], compareUsing compare: (T, T) -> Bool = (<)) -> [T] {
	var array = array
	let count	= array.count

	for _ in 0..<count {
		var swapped = false

		for innerIndex in 0..<count - 1 {
			let a = array[innerIndex]
			let b = array[innerIndex + 1]
			if compare(b, a) {
				array[innerIndex] = b
				array[innerIndex + 1] = a

				swapped = true
			}
		}

		if !swapped { break }
	}

	return array
}

File: BubbleSort/Java/Sources/BubbleSort/BubbleSort.java
//
//  BubbleSort.java
//  BubbleSort
//
//  Created by Ardalan Samimi on 2019-03-29.
//
package BubbleSort;

public class BubbleSort {

		public static void sort(int[] array) {
			int size = array.length;

			for (int i = 0; i < size; i++) {
				for (int j = 1; j < size - i; j++) {
					if (array[j - 1] > array[j]) {
						BubbleSort.swap(array, j, j - 1);
					}
				}
			}
		}

		private static void swap(int[] array, int i, int j) {
			int temp = array[i];
			array[i] = array[j];
			array[j] = temp;
		}

}
